feat: add GetEnvironments method to DBus register

rverdile opened the pull request:
Card-ID: CCT-764

This implements the functionality requested in the card, but I would like to get guidance on how to architect the implementation. Perhaps I should be implementing an environments class? The `get_environments` method I created makes multiple calls to the candlepin API, which I don't see precedence for looking at the other DBus methods.

rverdile commented:
tested using this script, a modified example from: https://github.com/jirihnidek/rhsm-dbus-examples/tree/main

```
#!/bin/bash

source ./library.sh

start_register_server
export username="admin"
export password="admin"
export org_id="content-sources-test"



echo "Trying to get list of organizations using dbus-send..."
dbus-send --address=${my_addr} --print-reply --dest='com.redhat.RHSM1.Register' \
       '/com/redhat/RHSM1/Register' \
       com.redhat.RHSM1.Register.GetEnvironments string:${username} \
       string:${password} \
       string:${org_id} \
       dict:string:string:"","" \
       string:""


stop_register_server
```

ptoscano commented on src/rhsm/connection.py:
```diff
@@ -2055,14 +2055,14 @@ def getServiceLevelList(self, owner_key: str) -> List[str]:
         can always check with supports_resource("environments").
         """
-        method = "/owners/%s/environments" % self.sanitize(owner_key)
+        method = "/owners/%s/environments?list_all=%r" % (self.sanitize(owner_key), list_all)
```
be careful here: we should not try to use bits that are not supported (ie the new `list_all` query parameter)

rverdile replied:
My understanding is we need to list_all for templates, but I updated this to only use this query parameter if typed environments are an available capability. Is that okay?

ptoscano reviewed (commented):
Thanks for this PR, and sorry for the wait.

I added a couple of notes of things I noticed so far; will give a better try later on.

jirihnidek commented on test/rhsmlib/dbus/test_register.py:
```diff
@@ -123,6 +123,56 @@
+    "type": "content-template"
+  }
+]
+"""
```
Please add more environments to this list. You will see that new unit test will start to fail.

rverdile replied:
I added more environments to this list (10 more), but I don't see an issue. If you're seeing something else, what else can I do to reproduce?

jirihnidek replied:
If the old code was used, then the unit test would fail.

jirihnidek reviewed (changes requested):
Thanks for the PR. I have few requests.

rverdile commented on src/rhsmlib/dbus/objects/register.py:
```diff
@@ -181,6 +181,32 @@ def get_organizations(self, options: dict) -> List[dict]:
+
+        list_all_environments: bool = False
+        if "typed_environments" in status["managerCapabilities"]:
+            list_all_environments = True
```
cool, thanks :)

rverdile commented on src/rhsmlib/dbus/objects/register.py:
```diff
@@ -181,6 +181,32 @@ def get_organizations(self, options: dict) -> List[dict]:
+
+        list_all_environments: bool = False
+        if "typed_environments" in status["managerCapabilities"]:
+            list_all_environments = True
```
thanks, added these

rverdile commented on src/rhsmlib/dbus/objects/register.py:
```diff
@@ -181,6 +181,32 @@ def get_organizations(self, options: dict) -> List[dict]:
         owners: List[dict] = uep.getOwnerList(options["username"])
         return owners
 
+    def get_environments(self, options: dict) -> List[dict]:
```
My understanding of the card is that the GetEnvironments() method is intended to be a DBus method. With that said, I moved the actual list call to candlepin to a newly created EnvironmentService class. I then instantiated that in the DBus method and used it to call to candlepin. Is that okay?

rverdile replied:
If I'm misunderstanding your request, I'm happy to make more changes :)

rverdile commented on src/rhsmlib/dbus/objects/register.py:
```diff
@@ -181,6 +181,32 @@ def get_organizations(self, options: dict) -> List[dict]:
+            env["name"] = e["name"]
+            env["description"] = e["description"]
+            env["type"] = e["type"]
+        environments.append(env)
```
One of the requirements of the card was to return a subset of the information (ID, name, type, description) from the DBus method

jirihnidek replied:
This requirement was wrong, please do not filter information in returned environments.

rverdile commented on src/rhsmlib/dbus/objects/register.py:
```diff
@@ -333,6 +359,40 @@ def __init__(self, conn=None, object_path=None, bus_name=None, sender=None, cmd_
+            dbus_sender.set_cmd_line(sender=self.sender, cmd_line=self.cmd_line)
+            Locale.set(locale)
+
+            owners = self.impl.get_environments(connection_options)
```
whoops :)..thanks

rverdile commented:
Thank you guys for the reviews! I addressed most of the feedback, and left some questions where I'm not sure. 

My team usually squashes commits before merging, so I just added a new commit for the feedback. Please let me know if I should organize my commits differently :)

rverdile commented:
updated, thanks for the review!

ptoscano commented on src/rhsmlib/services/environment.py:
```diff
@@ -0,0 +1,52 @@
+# Copyright (c) 2017 Red Hat, Inc.
```
I guess this should be this year?

ptoscano commented on src/rhsmlib/services/environment.py:
```diff
@@ -0,0 +1,52 @@
+        self.plugin_manager = inj.require(inj.PLUGIN_MANAGER)
+        self.cp = cp
+
+    def list(self, org_id: str, typed_environments: bool = True) -> List[dict]:
```
The `typed_environments` parameter is always true, so the "false" code path is effectively unused here. Since I think querying in the "best way" (i.e. with `list_all`) is generally a good idea regardless of what it is being done, I'd simply drop this parameter, and always use `list_all` when available/usable.

ptoscano commented on test/rhsmlib/services/test_environment.py:
```diff
@@ -0,0 +1,83 @@
+# Copyright (c) 2017 Red Hat, Inc.
```
ditto (date)?

ptoscano commented on src/rhsmlib/dbus/objects/register.py:
```diff
@@ -181,6 +182,33 @@ def get_organizations(self, options: dict) -> List[dict]:
+
+        environments = environment_service.list(options["org_id"], typed_environments=True)
+
+        keys_to_remove: List[str] = [
```
Minor nitpick: since this is more a constant than a variable, its name ought to be UPPER_CASE; I'd also move it as module-level variable with a more specific name, e.g. `ENVIRONMENTS_KEY_TO_FILTER`.

ptoscano commented on src/rhsmlib/dbus/objects/register.py:
```diff
@@ -181,6 +182,33 @@ def get_organizations(self, options: dict) -> List[dict]:
+        environments = environment_service.list(options["org_id"], typed_environments=True)
+
+        keys_to_remove: List[str] = [
+            "created",
+            "updated",
+            "contentPrefix",
+            "owner",
+            "environmentContent",
```
last nitpick of the round :) would you please alpha-sort this list, so it's easier to scan visually?

rverdile commented:
updated, thanks!

ptoscano reviewed (approved):
Thanks for the changes! The code now LGTM, and it works well too.

Since @jirihnidek had notes/reviews, I'll approve leaving it to him for an additional review.

One small last thing from me (a forgotten previous comment from you, sorry):

> My team usually squashes commits before merging, so I just added a new commit for the feedback. Please let me know if I should organize my commits differently :)

We generally don't squash commits on merging; would you please squash them in this PR, and provide a commit message that explains the changes done?

jirihnidek reviewed (approved):
LGTM, Thanks for updates
